### PR TITLE
Xi, Iota & Absorb gates

### PR DESCRIPTION
**Pull request by CPerezz**
Implementation of the Xi, Iota and Absorb Keccak circuits in Halo2 according to the spec  in https://hackmd.io/sK7v0lr8Txi1bgION1rRpw?view#Xi--Iota-step

### CODE REVIEW

File: keccak256/src/gates/absorb.rs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        AbsorbConfig {
+            q_enable,
+            state,
+            next_input,
+            _marker: PhantomData,
+        }
+    }
+    pub fn load(&self, _layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        Ok(())
+    }
```
Comment (therealyingtong) [inline keccak256/src/gates/absorb.rs:63]:
```suggestion
```

File: keccak256/src/gates/iota_b13.rs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            q_enable,
+            state,
+            round_ctant_b13,
+            round_constants,
+            _marker: PhantomData,
+        }
+    }
+    pub fn load(&self, _layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        Ok(())
+    }
```
Comment (therealyingtong) [inline keccak256/src/gates/iota_b13.rs:48]:
```suggestion
```

File: keccak256/src/gates/iota_b13.rs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                            self.in_state,
+                        )?;
+                        let offset = 1;
+                        config.assign_state(
+                            &mut region,
+                            offset,
+                            self.out_state,
+                        )?;
+                        let offset = 0;
+                        config.assign_round_ctant_b13(&mut region, offset, 0)
```
Comment (therealyingtong) [inline keccak256/src/gates/iota_b13.rs:164]:
```suggestion
                        config.assign_round_ctant_b13(&mut region, offset, self.round_ctant_b13)
```

File: keccak256/src/gates/iota_b9.rs [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+            q_enable,
+            state,
+            round_ctant_b9,
+            round_constants,
+            _marker: PhantomData,
+        }
+    }
+    pub fn load(&self, _layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        Ok(())
+    }
```
Comment (therealyingtong) [inline keccak256/src/gates/iota_b9.rs:52]:
```suggestion
```

File: keccak256/src/gates/iota_b9.rs [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                        let offset = 1;
+                        config.assign_state(
+                            &mut region,
+                            offset,
+                            self.out_state,
+                        )?;
+                        let offset = 0;
+                        // Within the Region itself, we use the constant in the same offset
+                        // so at position (Rotation::curr()). Therefore we use `0` here.
+                        config.assign_round_ctant_b9(&mut region, offset, 0)
```
Comment (therealyingtong) [inline keccak256/src/gates/iota_b9.rs:177]:
```suggestion
                        config.assign_round_ctant_b9(&mut region, offset, self.round_ctant_b9)
```

File: keccak256/src/gates/xi.rs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+        XiConfig {
+            q_enable,
+            state,
+            _marker: PhantomData,
+        }
+    }
+    pub fn load(&self, _layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        Ok(())
+    }
```
Comment (therealyingtong) [inline keccak256/src/gates/xi.rs:71]:
```suggestion
```

**Review by han0110 — COMMENTED** [review]
Nice work! Got some naive questions:

File: keccak256/src/gates/absorb.rs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    _marker: PhantomData<F>,
+}
+
+impl<F: FieldExt> AbsorbConfig<F> {
+    // We assume state is recieved in base-9.
+    pub fn configure(
+        q_enable: Selector,
+        meta: &mut ConstraintSystem<F>,
+        state: [Column<Advice>; 25],
+        next_input: [Column<Advice>; 25],
```
[the remark this answers is not in the document]
Comment (ChihChengLiang) [reply keccak256/src/gates/absorb.rs:24]: I have another question for `input` columns too. As I learned recently, the custom gates work on rows that are close to each other, like Rotation::cur or Rotation::next. If the input comes from somewhere, we'll have to do a copy constraint on them.

We probably need to pass the input cells in the `assign_state` function, and use `region.constrain_equal` to assign them to states.

File: keccak256/src/gates/iota_b13.rs [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+};
+use pasta_curves::arithmetic::FieldExt;
+use std::marker::PhantomData;
+
+#[derive(Clone, Debug)]
+pub struct IotaB13Config<F> {
+    q_enable: Selector,
+    state: [Column<Advice>; 25],
+    round_ctant_b13: Column<Advice>,
+    round_constants: Column<Instance>,
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply keccak256/src/gates/iota_b13.rs:15]: Constants would be hardcoded directly into the circuit. 
That means that if the next keccak instance uses less round constants, the circuit will be different.

In that case, using them as instance, and using copy constraints we can manage all the situations with the same circuit description and the same constraint number.

[the remark this answers is not in the document]
Comment (CPerezz) [reply keccak256/src/gates/iota_b13.rs:15]: Maybe I should add further docs on that. I thought I already did as @therealyingtong suggested.

File: keccak256/src/gates/absorb.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (CPerezz) [reply keccak256/src/gates/absorb.rs:24]: Ohh I see. Let me ammend the `next_input` length then. 
Then I guess we can merge and proceed with #105

File: keccak256/src/gates/iota_b13.rs [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+        state: [Column<Advice>; 25],
+        round_ctant_b13: Column<Advice>,
+        round_constants: Column<Instance>,
+    ) -> IotaB13Config<F> {
+        meta.create_gate("iota_b13", |meta| {
+            // def iota_b13(state: List[List[int], round_constant_base13: int):
+            // state[0][0] += round_constant_base13
+            // return state
+            let state_00 = meta.query_advice(state[0], Rotation::cur())
+                + meta.query_advice(round_ctant_b13, Rotation::cur());
```
Comment (ChihChengLiang) [inline keccak256/src/gates/iota_b13.rs:33]: Can we assign the `round_ctant_b13` as constants here?
```suggestion
                + Expression::Constant(F::from(round_ctant_b13));
```

**Review by ChihChengLiang — APPROVED** [review]
LGTM. There are still some other minor issues, but we might have more clarity when dealing with #105.

**Pull request merged**